light-clients/ics10-grandpa-cw/src/contract.rs

```diff
@@ -0,0 +1,361 @@
+				ClientMessage::Header(header) => {
+					use finality_grandpa::Chain;
+					let ancestry = AncestryChain::<RelayChainHeader>::new(
+						&header.finality_proof.unknown_headers,
+					);
+					let from = msg.client_state.latest_relay_hash;
+					let finalized =
+						ancestry.ancestry(from, header.finality_proof.block).map_err(|_| {
+							ContractError::Grandpa(format!("[update_state] Invalid ancestry!"))
+						})?;
+					finalized
+				},
+				_ => Vec::new(),
+			};
+
+			client
+				.update_state(ctx, client_id.clone(), msg.client_state, msg.client_message)
+				.map_err(|e| ContractError::Grandpa(e.to_string()))
+				.and_then(|(cs, cu)| {
+					let height = cs.latest_height();
+					client_state.latest_height = height.into();
+					client_state.data = cs.to_any().encode_to_vec();
+					ctx.insert_relay_header_hashes(&finalized_headers);
+
+					match cu {
+						ConsensusUpdateResult::Single(cs) => {
+							log!(ctx, "Storing consensus state: {:?}", height);
+							ctx.store_consensus_state(client_id.clone(), height, cs)
+								.map_err(|e| ContractError::Grandpa(e.to_string()))?;
+						},
+						ConsensusUpdateResult::Batch(css) =>
+							for (height, cs) in css {
+								log!(ctx, "Storing consensus state: {:?}", height);
+								ctx.store_consensus_state(client_id.clone(), height, cs)
+									.map_err(|e| ContractError::Grandpa(e.to_string()))?;
+							},
+					}
+					Ok(to_binary(&client_state)
+						.and_then(|data| to_binary(&ContractResult::success().data(data.0))))
+				})
+		},
+		ExecuteMsg::CheckSubstituteAndUpdateState(_msg) => {
+			todo!("check substitute and update state")
+		},
+		ExecuteMsg::VerifyUpgradeAndUpdateState(msg) => {
+			let msg = VerifyUpgradeAndUpdateStateMsg::try_from(msg)?;
+			client
+				.verify_upgrade_and_update_state(
+					ctx,
+					client_id,
+					&msg.old_client_state,
+					&msg.upgrade_client_state,
+					&msg.upgrade_consensus_state,
+					msg.proof_upgrade_client,
+					msg.proof_upgrade_consensus_state,
+				)
+				.map_err(|e| ContractError::Grandpa(e.to_string()))
+				.map(|_| to_binary(&ContractResult::success()))
+		},
+		ExecuteMsg::InitializeState(InitializeState { client_state, consensus_state }) => {
+			let state_call_response = ClientStateCallResponse {
+				new_consensus_state: consensus_state.clone(),
+				new_client_state: client_state.clone(),
+				client_state,
+				result: ContractResult::success(),
+			};
+			let response = to_binary(&state_call_response);
+			Ok(response)
+		},
+		ExecuteMsg::ClientCreateRequest(_) => Ok(to_binary(&ContractResult::success())),
+	};
+	Ok(result??)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(_deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+	match msg {
+		QueryMsg::ClientTypeMsg(_) => todo!(),
+		QueryMsg::GetLatestHeightsMsg(_) => todo!(),
+	}
+}
+
+#[no_mangle]
+// &[u8] -> [u8; 32]
+pub extern "C" fn ext_hashing_blake2_256_version_1(data: i64) -> i32 {
```

Comment:
what's the reason behind using FFI here?

Reply:
We're using them because of sp-io dependency that expects the functions to be available on the host.